### PR TITLE
Global Styles: Allow non-admin users access to global styles data in post editor

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
⚠️ Do not merge! ⚠️  

Related: https://github.com/WordPress/gutenberg/issues/64755

## What?

This draft PR is only intended to illustrate the sort of changes required to make global styles data available in the post editor to non-admin users (users without `edit_theme_options`).

## Why?

Global styles data is relied upon in the post editor to generate block style variation styles. It will also be needed in the current work around style inheritance and displaying inherited values in the block inspector controls.

## How?

- Allows users with `edit_posts` capabilities to read base theme global styles.
- Tries updating globalStyles entity definition to avoid enforcing edit context. (this is mostly just wild speculation and hacking due to lack of understanding)

## Testing Instructions
1. Create and log in with a non-admin user e.g. editor
2. Open the post editor, add a button, and assign it the outline block style, then save
3. Note that the styles aren't generated and applied in the editor
4. Checkout this PR and reload the editor
5. The outline style should now be applied
6. With an admin user, navigate to Appearance > Editor > Styles > Blocks > Button > Outline. Apply a custom background color for this block style variation.
7. Save the global styles changes and switch back to the non-admin user in the Post Editor
8. Reload the editor again and the customization made in global styles should be applied as well.

##### Known Issue

This approach seems to have reintroduced an issue in the post editor where there is a delay in block style variation styles being applied. This was solved previously by adding preloaded paths for global styles when in the post editor context. I'm not sure what I'm missing this time around.

### CODE REVIEW

**Comment by aaronrobertshaw** [comment]
@jorgefilipecosta or @youknowriad when you get a chance would you be able to point me in the right direction on how best to allow non-admin users access to global styles data in the post editor?

@talldan and I looked into this today but we could do with some expert opinions 🙏

File: packages/editor/src/components/global-styles-provider/index.js [hunk 1]
```diff
@@ -146,13 +146,10 @@ function useGlobalStylesUserConfig() {
 
 function useGlobalStylesBaseConfig() {
 	const baseConfig = useSelect( ( select ) => {
-		const { __experimentalGetCurrentThemeBaseGlobalStyles, canUser } =
+		const { __experimentalGetCurrentThemeBaseGlobalStyles } =
 			select( coreStore );
 
-		return (
-			canUser( 'read', { kind: 'root', name: 'theme' } ) &&
```
Comment (talldan) [inline packages/editor/src/components/global-styles-provider/index.js:153]: Something I just noticed, this is only checking whether the user can read the `theme`. While the `__experimentalGetCurrentThemeBaseGlobalStyles` does get the current theme as part of its logic, there's nothing to check whether the user can read the `globalStyles` entity (the actual entity type being returned). 🤔

Comment (aaronrobertshaw) [reply packages/editor/src/components/global-styles-provider/index.js:153]: How our permissions checks are supposed to work with entities is going over my head at the moment 🫠 

The REST API endpoint called when retrieving this base global styles data does check if the user has the `edit_posts` capability in this PR, or `edit_theme_options` on trunk. So is that not "something"?

Comment (talldan) [reply packages/editor/src/components/global-styles-provider/index.js:153]: When debugging I find it helpful to remove the preloading so that you can see the requests in your dev tools. The permissions checks from `canUser` are generally `OPTIONS` requests, and the `Allow` header in the response provides a list of the HTTP verbs that the current user is allowed to perform.

The `canUser` call you deleted here doesn't seem to work from what I can tell. I think that's because it doesn't match the actual call that's made to get the current theme. My understanding is that `canUser( 'read', { kind: 'root', name: 'theme' } )` is checking "can the user `GET` a list of installed themes" (which I'd expect lower tiered users not to be able to do). The actual request to get the current theme is querying for `status: active`:
https://github.com/WordPress/gutenberg/blob/482c18803ed9becccc57d1b23632438e1afc1d67/packages/core-data/src/resolvers.js#L374-L378

The `status: active` part changes the permission check in the REST Controller:
https://github.com/WordPress/wordpress-develop/blob/f4761a3f8c5aa6920d4407e3af271ef70db5f305/src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php#L103-L106

The user only needs `edit_posts` caps:
https://github.com/WordPress/wordpress-develop/blob/f4761a3f8c5aa6920d4407e3af271ef70db5f305/src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php#L149-L152

I'll have a look into whether there's a way to replace that with a check that actually works.

I think there's a similar issue with trying to do `canUser( 'read', { kind: 'root', name: 'globalStyles' } )`. That would check whether the user can list all global styles, but we actually need to check a route that's more like `globalStyles/theme/<activeStylesheet>`.

Comment (talldan) [reply packages/editor/src/components/global-styles-provider/index.js:153]: I've pushed a commit that adds working permissions checks.

Comment (aaronrobertshaw) [reply packages/editor/src/components/global-styles-provider/index.js:153]: So if the approach in https://github.com/WordPress/gutenberg/pull/64797/commits/790febdf67899a8d659dd894b5cadeec982f4135 is going back towards checking capabilities directly as before https://github.com/WordPress/gutenberg/pull/63812. Does that mean we can also roll back the changes to the global styles entity definition?

Comment (aaronrobertshaw) [reply packages/editor/src/components/global-styles-provider/index.js:153]: If pushing changes to this PR, it might also helps those pinged for input if the PR description is updated to match new changes.

File: packages/core-data/src/entities.js [hunk 2]
```diff
@@ -182,14 +182,25 @@ export const rootEntitiesConfig = [
 		name: 'globalStyles',
 		kind: 'root',
 		baseURL: '/wp/v2/global-styles',
-		baseURLParams: { context: 'edit' },
+		baseURLParams: {},
```
Comment (talldan) [inline packages/core-data/src/entities.js:185]: It seems to work ok with `context: edit`, so maybe it's ok to undo these changes.

File: packages/editor/src/components/global-styles-provider/index.js [hunk 3]
```diff
@@ -146,13 +146,30 @@ function useGlobalStylesUserConfig() {
...
 	const baseConfig = useSelect( ( select ) => {
-		const { __experimentalGetCurrentThemeBaseGlobalStyles, canUser } =
-			select( coreStore );
+		const {
+			__experimentalGetCurrentThemeBaseGlobalStyles,
+			getCurrentTheme,
+			canUser,
+		} = select( coreStore );
+
+		const canReadActiveTheme = canUser( 'read', 'themes?status=active' );
```
Comment (aaronrobertshaw) [inline packages/editor/src/components/global-styles-provider/index.js]: So to prevent a flash of global styles in the post editor, we'll need to preload these paths as well?

**Comment by aaronrobertshaw** [comment]
@talldan retesting this PR after the latest commits shows a regression. The global styles data isn't accessible to the non-admin user anymore and the block style variation styles aren't generated.

| https://github.com/WordPress/gutenberg/pull/64797/commits/06f1fbde8de7fdc26b0a5760b285790b0aa62f2e | [Latest](https://github.com/WordPress/gutenberg/pull/64797/commits/24d51e8c98bbd4bf361ac8576c01a54127baa1e1) |
|---|---|
| Screenshot 2024-08-27 at 4 12 08 PM | Screenshot 2024-08-27 at 4 12 29 PM |

In the before picture, the custom global styles for the button outline block style I set in Global Styles are applied. In the latest on this branch, now not even the base outline styles apply.

File: packages/editor/src/components/global-styles-provider/index.js (hunk 1 above)
Comment (talldan) [reply packages/editor/src/components/global-styles-provider/index.js:153]: Actually, seems like I was logged in as admin accidentally when I thought it was working. It doesn't work with non-admin so I'll revert those commits.

The explanation is still correct, but unfortunately it doesn't look like there's a way to check permissions using `canUser` for those requests.

**Comment by talldan** [comment]
> @talldan retesting this PR after the latest commits shows a regression. The global styles data isn't accessible to the non-admin user anymore and the block style variation styles aren't generated.

Yeah, my bad, I was switching users a lot and accidentally was logged in as admin when testing my changes. I've reverted those changes.

I'm not really sure of the best way to check permissions now if that doesn't work, perhaps it's best just to try making the request and if it fails, handle it.

File: packages/core-data/src/entities.js [hunk 4]
```diff
@@ -182,14 +182,25 @@ export const rootEntitiesConfig = [
...
 		plural: 'globalStylesVariations', // Should be different from name.
 		getTitle: ( record ) => record?.title?.rendered || record?.title,
 		getRevisionsUrl: ( parentId, revisionId ) =>
 			`/wp/v2/global-styles/${ parentId }/revisions${
 				revisionId ? '/' + revisionId : ''
 			}`,
 		supportsPagination: true,
+		getPath: ( path, query, baseURL, id ) => {
```
Comment (ramonjd) [inline packages/core-data/src/entities.js]: What does this do? I couldn't find where's it's called.

**Review by ramonjd — COMMENTED** [review]
Thanks for getting this PR started, and sorry if this is premature. Is the plan to allow read access to merged/global styles CPT data as well?

I just did some manual testing with various user roles to check what's available. In this PR the base theme styles are coming through 👍🏻 

```
// Check in the post editor for styles values:
await wp.data.resolveSelect( 'core/block-editor').getSettings() 

// Check read access
await wp.data.resolveSelect( 'core' ).getEntityRecord( 'root', 'globalStyles', await wp.data.resolveSelect( 'core' ).__experimentalGetCurrentGlobalStylesId() ) 
```

File: packages/core-data/src/entities.js (hunk 4 above)
Comment (aaronrobertshaw) [reply packages/core-data/src/entities.js]: Sorry, accidentally committed code I was playing with. I've removed it in https://github.com/WordPress/gutenberg/pull/64797/commits/a1b00fc5b92a43f3857f71f4468bb06d97e905b3.

I've left the removal of the `{ context: 'edit' }` in place to illustrate the merged global styles being retrieved for non-admins. This isn't a solution of course.

**Comment by aaronrobertshaw** [comment]
> Is the plan to allow read access to merged/global styles CPT data as well?

Yes, the complete merged global styles data is required in the post editor for non-admins to enable block style variation styles and style inheritance.

>  this PR the base theme styles are coming through 👍🏻

This is the first half of the equation. The part I'm less clear on how to tackle are the permissions for the global styles CPT entity storing the user set global styles.

This PR incorrectly removes the `context: edit` from the entity's base url params. This allows a non-admin user to retrieve the data but probably loosens edit restrictions that should remain in place.

**Comment by ramonjd** [comment]
>  The part I'm less clear on how to tackle are the permissions for the global styles CPT entity storing the user set global styles.

This might sound far-fetched - in my mind, a read-only global styles CPT belongs to the "site". 

It's owner (whichever admin starts to edit global styles) is a side effect of the fact that only admins can access global styles. 

So is it too simplistic to allow GET requests and do permissions checks on the rest?

Probably not yet relevant, but global styles revisions won't still only be readable by admins: https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/rest-api/endpoints/class-wp-rest-revisions-controller.php#L186

File: lib/class-wp-rest-global-styles-controller-gutenberg.php [hunk 5]
```diff
@@ -586,12 +586,11 @@ public function get_item_schema() {
...
 	public function get_theme_item_permissions_check( $request ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
-
 		/*
-		 * Verify if the current user has edit_theme_options capability.
-		 * This capability is required to edit/view/delete templates.
+		 * Verify if the current user has edit_posts capability.
+		 * This capability is required to view global styles.
 		 */
-		if ( ! current_user_can( 'edit_theme_options' ) ) {
+		if ( ! current_user_can( 'edit_posts' ) ) {
```
Comment (ramonjd) [inline lib/class-wp-rest-global-styles-controller-gutenberg.php:593]: To allow the `canUser( 'read', { kind: 'root', name: 'globalStyles', id: _globalStylesId, } )` check in `useGlobalStylesUserConfig()`  to pass we might need to update the read capabilities of the `wp_global_styles` custom post type:

https://github.com/ramonjd/wordpress-develop/blob/30523ee4b6a0d33b1c2bd8c9ca4021b6256a96c3/src/wp-includes/post.php#L492-L492

E.g., to `'read'                   => 'edit_posts',`

That would mean that editors/authors would be able to read the global styles post, e.g., `await wp.data.resolveSelect( 'core' ).getEntityRecord( 'root', 'globalStyles', await wp.data.resolveSelect( 'core' ).__experimentalGetCurrentGlobalStylesId() )`

It seems harmless enough, but I'm not sure what's kosher.

To display the values, the current editor/author needs to have read access to both.

I haven't thought this through, but I wonder if a new read-only endpoint to retrieve merged theme data might be in order. The `edit-post` package could call it directly from the core store and there'd be no need to shuffle around block settings.

**Comment by talldan** [comment]
I think it'd be good to bring some REST API folk into this conversation (maybe @TimothyBJacobs?). Also some other contributors that know global styles well (perhaps @oandregal  and @ajlende?).

(The TLDR is that as of 6.6 the post editor uses the global-styles endpoint, but a problem is that only admins (users with `edit_theme_options` caps) can access that endpoint, it results in users without that capability not being able to see some block styles.)

The proposal here is to lower the permissions to `edit_posts` for reading global styles. I see the themes endpoint does something similar (`check_read_active_theme_permission`). There's some nuance though as there's the user global styles and the theme global styles to consider.

File: lib/class-wp-rest-global-styles-controller-gutenberg.php (hunk 5 above)
Comment (youknowriad) [reply lib/class-wp-rest-global-styles-controller-gutenberg.php:593]: Yeah, I think if we think that we should allow "read" for global styles objects using the `edit_posts` capability, we should just make the update in the existing endpoint instead of creating a new one. 

It seems ok to me to do so. I don't see any harm for "editors" to have access to this information (without the ability to modify it)

Comment (TimothyBJacobs) [reply lib/class-wp-rest-global-styles-controller-gutenberg.php:593]: Just commenting on the REST API side of it, that makes sense to me. We do need to adapt the permission check to not just be `edit_posts` though, it needs to include any `show_in_rest` cpt. See `\WP_REST_Themes_Controller::check_read_active_theme_permission` for an example.

Comment (ramonjd) [reply lib/class-wp-rest-global-styles-controller-gutenberg.php:593]: Thanks for the feedback! 

Ah I see what you mean about [WP_REST_Themes_Controller::check_read_active_theme_permission](https://github.com/WordPress/wordpress-develop/blob/fb40fe915efab59d54d5bf812e2a4574ae525510/src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php#L149) - so if a user had rights to edit any type of post, they pass the test?

Comment (TimothyBJacobs) [reply lib/class-wp-rest-global-styles-controller-gutenberg.php:593]: Yep, exactly.

**Comment by ramonjd** [comment]
I've just been experimenting with the ideas proposed in this PR in:

- https://github.com/WordPress/gutenberg/pull/65071

Mainly to judge the approach, and test viability. There might be bugs 😄 but if folks think it's a reasonable direction we can pull it across to this PR.

**Comment by aaronrobertshaw** [comment]
Thanks for all the discussion and exploration here folks 🙇 

I'll close this PR in favour of https://github.com/WordPress/gutenberg/pull/65071 which looks promising. Thanks @ramonjd!
